src/IO/S3Common.cpp

```diff
@@ -85,15 +89,107 @@ class AWSLogger final : public Aws::Utils::Logging::LogSystemInterface
+                        absolute_uri, token.empty() ? "n empty" : " non-empty");
+            }
+            else if (Aws::Utils::StringUtils::ToLower(ec2_metadata_disabled.c_str()) != "true")
+            {
+                Aws::Client::ClientConfiguration aws_client_configuration;
+
+                /// See MakeDefaultHttpResourceClientConfiguration().
+                /// This is part of EC2 metadata client, but unfortunately it can't be accessed from outside
+                /// of contrib/aws/aws-cpp-sdk-core/source/internal/AWSHttpResourceClient.cpp
+                aws_client_configuration.maxConnections = 2;
+                aws_client_configuration.scheme = Aws::Http::Scheme::HTTP;
+
+                /// Explicitly set the proxy settings to empty/zero to avoid relying on defaults that could potentially change
+                /// in the future.
+                aws_client_configuration.proxyHost = "";
+                aws_client_configuration.proxyUserName = "";
+                aws_client_configuration.proxyPassword = "";
+                aws_client_configuration.proxyPort = 0;
+
+                /// EC2MetadataService throttles by delaying the response so the service client should set a large read timeout.
+                /// EC2MetadataService delay is in order of seconds so it only make sense to retry after a couple of seconds.
+                aws_client_configuration.connectTimeoutMs = 1000;
+                aws_client_configuration.requestTimeoutMs = 1000;
+                aws_client_configuration.retryStrategy = std::make_shared<Aws::Client::DefaultRetryStrategy>(1, 1000);
+
+                DB::S3::PocoHTTPClientConfiguration client_configuration(aws_client_configuration, remote_host_filter, s3_max_redirects);
+                auto ec2_metadata_client = std::make_shared<Aws::Internal::EC2MetadataClient>(client_configuration);
+                auto config_loader = std::make_shared<Aws::Config::EC2InstanceProfileConfigLoader>(ec2_metadata_client);
+
+                AddProvider(std::make_shared<Aws::Auth::InstanceProfileCredentialsProvider>(config_loader));
+                LOG_INFO(logger, "Added EC2 metadata service credentials provider to the provider chain.");
+            }
+        }
+
+        AddProvider(std::make_shared<Aws::Auth::SimpleAWSCredentialsProvider>(credentials));
+    }
+};
+
 class S3AuthSigner : public Aws::Client::AWSAuthV4Signer
 {
 public:
     S3AuthSigner(
         const Aws::Client::ClientConfiguration & client_configuration,
         const Aws::Auth::AWSCredentials & credentials,
-        const DB::HeaderCollection & headers_)
+        const DB::HeaderCollection & headers_,
+        bool use_environment_credentials)
         : Aws::Client::AWSAuthV4Signer(
-            std::make_shared<Aws::Auth::SimpleAWSCredentialsProvider>(credentials),
+            std::make_shared<S3CredentialsProviderChain>(
```

Comment:
This constructor would also become less coupled if it consumed `Aws::Auth::AWSCredentialsProvider`